zrml/asset-router/src/macros.rs

```diff
@@ -0,0 +1,98 @@
+    ($currency_id:expr, $currency_method:ident, $asset_method:ident, $($args:expr),*) => {
+        if let Ok(currency) = T::CurrencyType::try_from($currency_id) {
+            Ok(<T::Currencies as MultiCurrency<T::AccountId>>::$currency_method(currency, $($args),*))
+        } else if let Ok(asset) = T::MarketAssetType::try_from($currency_id) {
+            Ok(T::MarketAssets::$asset_method(asset, $($args),*))
+        } else if let Ok(asset) = T::CampaignAssetType::try_from($currency_id) {
+            Ok(T::CampaignAssets::$asset_method(asset, $($args),*))
+        } else if let Ok(asset) = T::CustomAssetType::try_from($currency_id)  {
+            Ok(T::CustomAssets::$asset_method(asset, $($args),*))
+        } else {
+            Err(Error::<T>::UnknownAsset)
+        }
+    };
+}
+
+/// This macro delegates a call to Currencies if the asset represents a currency, otherwise
+/// it returns an error.
+macro_rules! only_currency {
+    ($currency_id:expr, $error:expr, $currency_trait:ident, $currency_method:ident, $($args:expr),+) => {
+        if let Ok(currency) = T::CurrencyType::try_from($currency_id) {
+            <T::Currencies as $currency_trait<T::AccountId>>::$currency_method(currency, $($args),+)
+        } else {
+            Self::log_unsupported($currency_id, stringify!($currency_method));
+            $error
+        }
+    };
+}
+
+/// This macro delegates a call to one *Asset instance if the asset does not represent a currency, otherwise
+/// it returns an error.
+macro_rules! only_asset {
+    ($asset_id:expr, $error:expr, $asset_trait:ident, $asset_method:ident, $($args:expr),*) => {
+        if let Ok(asset) = T::MarketAssetType::try_from($asset_id) {
+            <T::MarketAssets as $asset_trait<T::AccountId>>::$asset_method(asset, $($args),*)
+        } else if let Ok(asset) = T::CampaignAssetType::try_from($asset_id) {
+            T::CampaignAssets::$asset_method(asset, $($args),*)
+        } else if let Ok(asset) = T::CustomAssetType::try_from($asset_id)  {
+            T::CustomAssets::$asset_method(asset, $($args),*)
+        } else {
+            Self::log_unsupported($asset_id, stringify!($asset_method));
+            $error
+        }
+    };
+}
+
+/// This macro handles the single stages of the asset destruction.
+macro_rules! handle_asset_destruction {
+    ($asset:expr, $remaining_weight:expr, $asset_storage:expr, $asset_method:ident, $outer_loop:tt) => {
+        let state_before = *($asset.state());
+        let call_result = Self::$asset_method($asset, $remaining_weight);
+        $remaining_weight = call_result.map_or_else(|weight| weight, |weight| weight);
+
+        if call_result.is_err() {
+            if state_before != DestructionState::Finalization {
+                break $outer_loop;
+            } else {
+                // In case destruction failed during finalization, there is most likely still
+                // some weight available.
+                break;
+            }
+        }
+
```

Comment:
The return value [is fixed](https://github.com/zeitgeistpm/zeitgeist/pull/1201/commits/58f7e1be1c15944a26f6a6d131c2503fea5965e0), but we still want that different break scenarios in the case an asset is indestructible, since there will be weight available when the previous `DestructionState` was `Finalization`, contrary to other states where we (now) play it safe and reduce the remaining weight to zero.